### PR TITLE
Refactor ScrollBar to utilise meta providers

### PR DESCRIPTION
**Pull request by kitsonk**
This migrates ScrollBar to using meta providers to provide it information about the DOM state.

### CODE REVIEW

**Comment by kitsonk** [comment]
Need to refactor again now that these are available in `widget-core`.

File: src/support/ScrollBar.ts [hunk 1]
```diff
@@ -99,83 +140,52 @@ export default class ScrollBar extends ThemeableBase<ScrollBarProperties> {
...
 		return v('div', {
-			classes: this.classes(css.root, css.vertical, visible ? css.visible : css.invisible),
+			classes: this.classes(
+				css.root,
+				horizontal ? css.horizontal : css.vertical,
+				visible || dragging ? css.visible : css.invisible
+			),
 			key,
 
+			onclick: this._onclick,
```
Comment (tomdye) [inline src/support/ScrollBar.ts]: `onpointerdown` ?

File: src/support/ScrollBar.ts [hunk 2]
```diff
@@ -51,38 +53,73 @@ const DEFAULT_KEY = 'scrollbar';
...
+			const domSize = this._getDomSize();
+			const {
+				horizontal,
+				position,
+				size = domSize,
+				sliderMin = 10,
+				sliderSize,
+				onScroll
+			} = this.properties;
+			const absoluteDelta = (horizontal ? evt.offsetX : evt.offsetY) -
```
Comment (tomdye) [inline src/support/ScrollBar.ts]: i don't follow why you need to go from relative delta to absolute and back to relative again here. Can't you just record the `start` position of the first `pointerdown` and scroll relative to there? I may be missing something.

File: src/support/ScrollBar.ts (hunk 1 above)
Comment (kitsonk) [reply src/support/ScrollBar.ts]: That wouldn't work/would be wrong.  `onpointerdown` is when the pointer is down, equivalent to `onmousedown` not `onclick`.  [There is no "pointer event" for "click"](https://www.w3.org/Submission/pointer-events/#extensions-to-the-element-interface).  Single taps are reported as `onclick`.

File: src/support/ScrollBar.ts (hunk 2 above)
Comment (kitsonk) [reply src/support/ScrollBar.ts]: The events works in absolute... so in order to calculate the delta.  The scroll bar only deals with relative values, so at some point you have to take the other information you need to calculate the delta, and convert it into absolute.  So this is taking an absolute offset and subtracting that from a relative internal state.  At some point, to compare apples and oranges, you need oranges and oranges, or apples and apples.

File: src/support/ScrollBar.ts [hunk 3]
```diff
@@ -51,38 +53,73 @@ const DEFAULT_KEY = 'scrollbar';
 const ThemeableBase = ThemeableMixin(WidgetBase);
 
 /**
+ * Convert a relative number to an absolute number
+ * @param relative The relative value to convert to an absolute value
+ * @param relativeSize The relative size to compare with the abosolute size
```
Comment (dylans) [inline src/support/ScrollBar.ts]: abosolute -> absolute

File: src/support/ScrollBar.ts [hunk 4]
```diff
@@ -51,38 +53,73 @@ const DEFAULT_KEY = 'scrollbar';
...
+ * @param absoluteSize The absolute size to compare with the relative size
+ */
+function fromRelative(relative: number, relativeSize: number, absoluteSize: number): number {
+	return absoluteSize * (relativeSize ? relative / relativeSize : 0);
+}
+
+/**
+ * Convert an absolute number to a relative number
+ * @param absolute The absolute value to convert to a relative value
+ * @param relativeSize The realtive size to compare with the absolute size
```
Comment (dylans) [inline src/support/ScrollBar.ts]: realtive -> relative

File: src/support/ScrollBar.ts [hunk 5]
```diff
@@ -51,38 +53,73 @@ const DEFAULT_KEY = 'scrollbar';
...
-
-	private _toRelative = (absolute: number): number => {
-		const size = this._getSize();
-		return this._domSize ? (absolute / this._domSize) * size : 0;
+	/**
+	 * Returns the height or width of the scroll bar, depending on if it is horizontal or not
+	 */
+	private _getDomSize(): number {
+		const { horizontal, key = DEFAULT_KEY } = this.properties;
+		// TODO: Remove string coersion when dojo/widget-core#721 is published
```
Comment (dylans) [inline src/support/ScrollBar.ts]: coersion -> coercion

File: src/support/ScrollBar.ts [hunk 6]
```diff
@@ -51,38 +53,73 @@ const DEFAULT_KEY = 'scrollbar';
...
-		const { horizontal = false, key: widgetKey = DEFAULT_KEY } = this.properties;
-		if (key === widgetKey) {
-			this._domSize = horizontal ? element.clientWidth : element.clientHeight;
+	/**
+	 * Determine if the scroll bar node is being clicked and if so, call the scroll listener with the relative position
+	 * that is attempted to be navigated to.
+	 * @param evt The mouse event
+	 */
+	private _onclick(evt: MouseEvent) {
+		// TODO: Remove string coersion when dojo/widget-core#721 is published
```
Comment (dylans) [inline src/support/ScrollBar.ts]: coersion -> coercion

**Pull request merged**